File: IPWorks SMIME Samples/Sign and Encrypt Email/java/ssmtp.java
/*
 * IPWorks S/MIME 2024 Java Edition - Sample Project
 *
 * This sample project demonstrates the usage of IPWorks S/MIME in a 
 * simple, straightforward way. It is not intended to be a complete 
 * application. Error handling and other checks are simplified for clarity.
 *
 * www.nsoftware.com/ipworkssmime
 *
 * This code is subject to the terms and conditions specified in the 
 * corresponding product license agreement which outlines the authorized 
 * usage and restrictions.
 */

import java.io.*;
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;
import ipworkssmime.*;

public class ssmtp {
  Vector al;
  //Vector<String> al;
  boolean quit;
  int index;
  SSMTP ssmtp1;
  CertMgr certmgr;
  public ssmtp() {
    al = new Vector();
    //al = new Vector<String>();
    index = 0;
    go();
  }

  public void go() {
    try {
      certmgr = new CertMgr();
      ssmtp1 = new SSMTP();
      certmgr.addCertMgrEventListener(new CertMgrEvents(this));
      ssmtp1.resetHeaders();
      ssmtp1.setTimeout(30);
      System.out.print("SMTP Server: ");
      ssmtp1.setMailServer(input());
      System.out.print("From: ");
      ssmtp1.setFrom(input());
      System.out.print("To: ");
      ssmtp1.setSendTo(input());
      System.out.print("Subject: ");
      ssmtp1.setSubject(input());
      ssmtp1.setMessageText("This is a test message from ssmtp.\r\n");
      quit = true;
      selectEncoding();
      System.out.print("Send? [Y/N]: ");
      char selection = input().toLowerCase().charAt(0);
      if (selection == 'y') {
        ssmtp1.send();
      }
      System.out.print("Would you like to send another? [Y/N]: ");
      selection = input().toLowerCase().charAt(0);
      if (selection == 'n') {
        return;
      }
      go();
    }
    catch (IPWorksSMIMEException ex) {
      System.out.println("IPWorksSMIMEException thrown: " + ex.getCode() + " [" +
                         ex.getMessage() + "].");
    }
    catch (Exception ex) {
      System.out.println(ex.getMessage());
    }
  }

  public void selectEncoding() throws IOException, IPWorksSMIMEException {
    char ch;
    System.out.println("\nEncoding choices:\n" +
                       "  1) Sign message\n" +
                       "  2) Encrypt message\n" +
                       "  3) Sign and encrypt message\n" +
                       "  0) Quit");
    ch = input().charAt(0);

    switch (ch) {
      case '1':
        sign();
        break;
      case '2':
        encrypt();
        break;
      case '3':
        signAndEncrypt();
        break;
      case '0':
        quit = true;
        break;
    }
  }

  private void sign() throws IPWorksSMIMEException {
    selectCertificate();
    ssmtp1.setCertificate(certmgr.getCert());
    ssmtp1.setIncludeCertificate(true);
    ssmtp1.sign();
  }

  private void encrypt() throws IPWorksSMIMEException {
    selectCertificate();
    ssmtp1.addRecipientCert(certmgr.getCert().getEncoded());
    ssmtp1.encrypt();
  }

  private void signAndEncrypt() throws IPWorksSMIMEException {
    System.out.println("Please select signing certificate:");
    selectCertificate();
    ssmtp1.setCertificate(certmgr.getCert());
    ssmtp1.setIncludeCertificate(true);
    System.out.println("Please select encrypting certificate:");
    selectCertificate();
    ssmtp1.addRecipientCert(certmgr.getCert().getEncoded());
    ssmtp1.signAndEncrypt();
  }

  private String input() throws IOException {
    BufferedReader bf = new BufferedReader(new InputStreamReader(System.in));
    return bf.readLine();
  }

  public static void main(String[] args) {
	new ssmtp();
  }

  private void selectCertificate() {
    try {
      String buffer;
      certmgr.setCertStoreType(Certificate.cstJKSFile);
      System.out.print("Please enter java key store path [myidentities.jks]: ");
      buffer = input();
      if ( buffer.length()==0 ) {
        buffer = "myidentities.jks";
      }
      certmgr.setCertStore(buffer);
      System.out.print("Please enter store password [password]: ");
      buffer = input();
      if ( buffer.length() == 0 ) {
        buffer = "password";
      }
      certmgr.setCertStorePassword(buffer);
      System.out.println("Please select a certificate:");
      certmgr.listStoreCertificates();
      int selection = Integer.parseInt(input());
      certmgr.setCert(new Certificate(
          Certificate.cstJKSFile,
          certmgr.getCertStore(),
          certmgr.getCertStorePassword(),
          (String) al.get(selection - 1) // noneed for cast if you are using 1.5 generics
                      ));

    }
    catch (IPWorksSMIMEException ex) {
      System.out.println("IPWorksSMIMEException thrown: " + ex.getCode() + " [" +
                         ex.getMessage() + "].");
    }
    catch (Exception ex) {
      System.out.println(ex.getMessage());
    }
  }

  public void certList(CertMgrCertListEvent args) {
    al.add(args.certSubject); //jdk1.5 should throw a warning here because of typed vectors in 1.5 (see vector definitons above)
    index++;
    System.out.println(index + ". " + args.certSubject);
  }

  public void error(CertMgrErrorEvent args) {
    System.out.println("CertMgr error: " + args.errorCode + " [" +
                       args.description + "].");
  }
}

class CertMgrEvents implements CertMgrEventListener{
		ssmtp instance;
  public CertMgrEvents(ssmtp instance) {
    this.instance = instance;
  }

  public void certChain(CertMgrCertChainEvent args) {

  }

  public void certList(CertMgrCertListEvent args) {
    instance.certList(args);
  }

  public void error(CertMgrErrorEvent args) {
    instance.error(args);
  }

  public void keyList(CertMgrKeyListEvent args) {

  }

  public void storeList(CertMgrStoreListEvent args) {

  }
  
  public void log(CertMgrLogEvent args){}
}

class ConsoleDemo {
  private static BufferedReader bf = new BufferedReader(new InputStreamReader(System.in));

  static String input() {
    try {
      return bf.readLine();
    } catch (IOException ioe) {
      return "";
    }
  }
  static char read() {
    return input().charAt(0);
  }

  static String prompt(String label) {
    return prompt(label, ":");
  }
  static String prompt(String label, String punctuation) {
    System.out.print(label + punctuation + " ");
    return input();
  }
  static String prompt(String label, String punctuation, String defaultVal) {
      System.out.print(label + " [" + defaultVal + "]" + punctuation + " ");
      String response = input();
      if (response.equals(""))
        return defaultVal;
      else
        return response;
  }

  static char ask(String label) {
    return ask(label, "?");
  }
  static char ask(String label, String punctuation) {
    return ask(label, punctuation, "(y/n)");
  }
  static char ask(String label, String punctuation, String answers) {
    System.out.print(label + punctuation + " " + answers + " ");
    return Character.toLowerCase(read());
  }

  static void displayError(Exception e) {
    System.out.print("Error");
    if (e instanceof IPWorksSMIMEException) {
      System.out.print(" (" + ((IPWorksSMIMEException) e).getCode() + ")");
    }
    System.out.println(": " + e.getMessage());
    e.printStackTrace();
  }

  /**
   * Takes a list of switch arguments or name-value arguments and turns it into a map.
   */
  static java.util.Map<String, String> parseArgs(String[] args) {
    java.util.Map<String, String> map = new java.util.HashMap<String, String>();
    
    for (int i = 0; i < args.length; i++) {
      // Add a key to the map for each argument.
      if (args[i].startsWith("-")) {
        // If the next argument does NOT start with a "-" then it is a value.
        if (i + 1 < args.length && !args[i + 1].startsWith("-")) {
          // Save the value and skip the next entry in the list of arguments.
          map.put(args[i].toLowerCase().replaceFirst("^-+", ""), args[i + 1]);
          i++;
        } else {
          // If the next argument starts with a "-", then we assume the current one is a switch.
          map.put(args[i].toLowerCase().replaceFirst("^-+", ""), "");
        }
      } else {
        // If the argument does not start with a "-", store the argument based on the index.
        map.put(Integer.toString(i), args[i].toLowerCase());
      }
    }
    return map;
  }
}

File: IPWorks SMIME Samples/Sign and Encrypt Email/c++/ssmtp.cpp
/*
 * IPWorks S/MIME 2024 C++ Edition - Sample Project
 *
 * This sample project demonstrates the usage of IPWorks S/MIME in a 
 * simple, straightforward way. It is not intended to be a complete 
 * application. Error handling and other checks are simplified for clarity.
 *
 * www.nsoftware.com/ipworkssmime
 *
 * This code is subject to the terms and conditions specified in the 
 * corresponding product license agreement which outlines the authorized 
 * usage and restrictions.
 */


#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "../../include/ipworkssmime.h"
#define LINE_LEN 100

#define MAX_CERTS 100
#define MAX_SUBJECT_LEN 499
class MyCertMgr : public CertMgr
{
public:

	MyCertMgr()
	{
		numCerts = 0;
	}
	virtual int FireCertList(CertMgrCertListEventParams *e)
	{
		if (numCerts < MAX_CERTS)
		{
			strncpy(certs[numCerts], e->CertSubject, MAX_SUBJECT_LEN);
			certs[numCerts][MAX_SUBJECT_LEN] = '\0';
			numCerts++;
		}
		return 0;
	}
	virtual int FireError(CertMgrErrorEventParams *e)
	{
		printf ("%i: %s\n", e->ErrorCode, e->Description);
		return 0;
	}
	char certs[MAX_CERTS][MAX_SUBJECT_LEN+1]; // storage for 100 cert subjects
	int numCerts;														// number of cert subjects stored in certs[]
};

SSMTP			ssmtp;		// global
MyCertMgr CertMgr1;	// global

int SelCert()
{

	char myBuf[LINE_LEN+1];
	*myBuf = 0;
	fflush(stdin);

	CertMgr1.numCerts = 0;
	CertMgr1.SetCertStore("MY", 2);
	CertMgr1.ListStoreCertificates();

	for (int i=0; i<CertMgr1.numCerts; i++)
	{
		printf("%i) %s\n", i, CertMgr1.certs[i]);
	}
	printf("Please select a valid certificate. ");
	fgets(myBuf,LINE_LEN,stdin);
	myBuf[strlen(myBuf)-1] = '\0';

	CertMgr1.SetCertSubject(CertMgr1.certs[atoi(myBuf)]);
	return 0;
}

int Sign()
{
	char* certStore;
	int certStoreLength;
	SelCert();
	//ssmtp.SetCertHandle(CertMgr1.GetCertHandle());
	CertMgr1.GetCertStore(certStore, certStoreLength);
	ssmtp.SetCertStore(certStore, certStoreLength);
	ssmtp.SetCertSubject(CertMgr1.GetCertSubject());
	ssmtp.SetIncludeCertificate(true);
	ssmtp.Sign();

	return 0;
}

int Encrypt()
{

	char* mycert;
	unsigned int   len = 0;
	SelCert();
	CertMgr1.GetCertEncoded(mycert, (int &)len);
	ssmtp.AddRecipientCert(mycert, len);
	ssmtp.Encrypt();

	return 0;
}

int SignAndEncrypt()
{

	char* certStore;
	int certStoreLength;
	char * mycert;
	unsigned int len = 0;
	SelCert();															// cert to sign message with
	ssmtp.GetCertStore(certStore, certStoreLength);
	ssmtp.SetCertSubject(CertMgr1.GetCertSubject());
	ssmtp.SetIncludeCertificate(true);
	printf("\n");
	SelCert();															// cert to encrypt message with
	CertMgr1.GetCertEncoded(mycert, (int &)len);
	ssmtp.AddRecipientCert(mycert, len);
	ssmtp.SignAndEncrypt();

	return 0;
}

int SelectEncoding()
{

	char ch;

	printf("\nEncoding choices:\n"
	       "  1) Sign message\n"
	       "  2) Encrypt message\n"
	       "  3) Sign and encrypt message\n"
	       "  4) No encryption\n");
	ch = getchar();

	switch(ch)
	{
		case '1':
			return Sign();
		case '2':
			return Encrypt();
		case '3':
			return SignAndEncrypt();
		case '4':
			return 0;
		default:
			break;
	}

	return 0;
}


int main()
{

	char buffer[LINE_LEN+1];
	char messageText[10000];
	*messageText = 0;

	ssmtp.ResetHeaders();
	ssmtp.SetTimeout(30);

	printf("SMTP Server: ");
	fgets(buffer,LINE_LEN,stdin);
	buffer[strlen(buffer)-1] = '\0';
	ssmtp.SetMailServer(buffer);

	printf("From:        ");
	fgets(buffer,LINE_LEN,stdin);
	buffer[strlen(buffer)-1] = '\0';
	ssmtp.SetFrom(buffer);

	printf("To:          ");
	fgets(buffer,LINE_LEN,stdin);
	buffer[strlen(buffer)-1] = '\0';
	ssmtp.SetSendTo(buffer);

	printf("Subject:     ");
	fgets(buffer,LINE_LEN,stdin);
	buffer[strlen(buffer)-1] = '\0';
	ssmtp.SetSubject(buffer);

	*buffer = 0;
	printf("\nBegin your message text. (type '.' on a line by itself to end)\n");

	while(strcmp(buffer, ".") != 0)
	{
		fgets(buffer,LINE_LEN,stdin);
		buffer[strlen(buffer)-1] = '\0';
		strcat(messageText, buffer);
	}
	ssmtp.SetMessageText(messageText);
	SelectEncoding();
	printf("\n%s\n", ssmtp.GetMessageText());
	fflush(stdin);
	printf("\nSend? [Y/N] ");
	fgets(buffer,LINE_LEN,stdin);
	buffer[strlen(buffer)-1] = '\0';
	if (*buffer == 'Y' || *buffer == 'y')
	{
		int ret_code = ssmtp.Send();
		if (ret_code == 0) printf("Message sent.\n\n");
		else printf("%i: %s\n\n", ret_code, ssmtp.GetLastError());
	}

}


